src/Graph/DOT.elm

```diff
@@ -125,13 +120,14 @@ outputWithStylesAndAttributes styles nodeAttrs edgeAttrs graph =
         attrAssocs : Dict String String -> String
         attrAssocs =
             Dict.toList
-                >> List.map (\( k, v ) -> k ++ "=" ++ Basics.toString v)
+                >> List.map (\( k, v ) -> k ++ "=\"" ++ v ++ "\"")
```

Comment:
So people on elm-slack gave me this to escape Strings:
`Json.Encode.encode 0 (Json.Encode.string "I say \"hi\"!")`
Not sure it is worth the trouble to incur dependency on `elm\json` package though.

An alternative would be to use elm parser:
```elm
Parser.loop "" (\soFar ->
  Parser.oneOf
    [ Parser.end |> Parser.map (\_ -> Parser.Done soFar)
    , Parser.token "\\" |> Parser.map (\_ -> Parser.Loop (soFar ++ "\\\\"))
    , Parser.token "\"" |> Parser.map (\_ -> Parser.Loop (soFar ++ "\\\""))
    , Parser.chompWhile (c -> c /= '\\' && c /= '"')
        |> Parser.getChompedString
        |> Parser.map (\newStuff -> (Parser.Loop (soFar ++ newStuff))
    ])
```

Reply:
So I looked at the [DOT language specification](https://graphviz.gitlab.io/_pages/doc/info/lang.html) and it has this to say about escaping Strings:

> In quoted strings in DOT, the only escaped character is double-quote ("). That is, in quoted strings, the dyad \" is converted to "; all other characters are left unchanged. In particular, \\ remains \\. Layout engines may apply additional escape sequences.

So the solution would be trivial - just use something like
```elm
escapeDotString :: String -> String
escapeDotString = String.replace "\"" "\\\""
```
at all places where you're rendering Strings to dot file.

Reply:
I started writing some tests for this, but ended up scrapping it for the following reasons:
- I literally cannot write the test without escaping things correctly (`"this has a \" and a \\ in it"`) or it fails to compile, and when the string is written like that, the tests would work as expected as is without the code to escape it
- I cannot write the expected cases (using multiline strings) with the escaping characters in place, because the formatter removes them, and if I use normal strings then I am just comparing two literals that are already the same
- The first two arguments to `Graph.DOT.output` are `n -> Maybe String` and `e -> Maybe String`, so the client should already be taking care of escaping strings in those functions, or ensuring they pass valid DOT strings in to begin with
- A similar thing applies to `outputWithStylesAndAttributes` - the client supplies the conversion to strings, we just need to add the DOT "glue" between those things, and the documentation say that.

(Note, when I tested this in https://dreampuf.github.io/GraphvizOnline/ I needed to escape backslashes as well as double quotes for it to compile.)

Reply:
Seems like that site isn't really adhering to the DOT standard.

> I literally cannot write the test without escaping things correctly ("this has a \" and a \\ in it") or it fails to compile, and when the string is written like that, the tests would work as expected as is without the code to escape it

I don't think your code above would format the string `"a\"b"` correctly as `"\"a\\\"b\""`. You should at least escape double quotes (for example by using `String.replace`, as above). I'd say we take the slight incompatability with that GraphViz of not escaping back slashes, so it's really just `"=\"" ++ String.replace "\"" "\\\"" v ++ "\""`.

> I cannot write the expected cases (using multiline strings) with the escaping characters in place, because the formatter removes them, and if I use normal strings then I am just comparing two literals that are already the same

Sorry, I don't follow here. Isn't that exactly what the test suite was supposed to catch? An example would help.

> the client should already be taking care of escaping strings

I don't think that the client should be bothered with knowing how to escape strings in DOT files. That's what this library is for. You would expect the same from a JSON API.

> we just need to add the DOT "glue" between those things, and the documentation say that.

I think you're referring to this part of the documentation:

> Note that you have to take care of setting the appropriate node and edge labels
yourself.

I'm not the author of that module, but I suppose this means you have to set the `node` and `edge` labels, not that you have to escape the values they are set to.

Regardless, I'll just take it from here.

Reply:
Ok, take a poke at it.  Just to clarify some points:

* If I gave a label to a node that looked  like `Node 0 "a\"b"` and ran it through the output code (as it is), and copy/paste the result into graphViz, the label on the rendered node would be `a"b` (the DOT string for that would be `label = "a\"b";`), which is what I would expect.  If I wanted the rendered label to be `a\"b` (and I don't know why I would want that) I would have to supply a label of `Node 0 "a\\\"b"` (for Elm to compile), which would turn out as expected, both in rendered graphViz and DOT.  
  In other words, since this is not valid Elm code: `Node 0 "a"b"`, I am forced to properly escape the strings for elm, which pass through to the DOT output.  *My main concern is that if we add a layer of escaping, if I made a `Node 0 "a\"b"` and the output was `label = "a\\\"b";` which renders as `a\"b` that would not be what I expect.*  So be sure to test that end to end (with https://dreampuf.github.io/GraphvizOnline/)

* After elm-format, `"""a\"b"""` becomes `"""a"b"""`, so I can't test that escaping is preserved if my expectations are in multiline strings.

I'm actually the original author of the DOT module (#8) though it has been modified since.

**Caveat: I have been looking at the output in the terminal when the tests run, and I think that has an extra layer of escaping done to it before printing it, so that may be having an effect.  When I have used this library, I have used the DOT in 3 ways - either by logging it in the console and copy/pasting, or rendering it into a textarea (or other element) to copy/paste, or passing it directly to graphViz (see https://github.com/jschomay/ene-graph/blob/master/src/graph.js).  I wouldn't be surprised if the escaping comes out differently between those 3 methods.  I haven't been able to test those 3 approaches now, but we definitely should.  I think having it work by passing it to the graphViz library is the most important behavior to have working (though for debugging I usually copy it out of the console).  Can you make an example app similar to the one I linked to to fully test this out before finalizing the approach?**

Let me know if you have any other questions.

Reply:
> the label on the rendered node would be a"b

Are you sure about this? Here is a REPL session:

```Elm
> import Graph
> import Graph.DOT
> g = Graph.fromNodeLabelsAndEdgePairs [ "a\"b" ] []
Graph (Leaf { key = 0, value = { incoming = Empty, node = { id = 0, label = "a\"b" }, outgoing = Empty } })
    : Graph.Graph String ()
> Graph.DOT.output Just (always Nothing) g
"digraph G {\n  rankdir=TB\n  graph []\n  node []\n  edge []\n\n\n\n  0 [label=\"a\"b\"]\n}"
    : String
```

This is not what I would expect. Note that the REPL itself escapes the string here, so really the plain `String` returned is

```
digraph G {
  rankdir=TB
  graph []
  node []
  edge []



  0 [label="a"b"]
}
```

Paste this in the DOT formatter of your choice and be convinced this is not right. Granted, this means you can't just copy and paste from the REPL, but that's just a weakness of Elm's lacking console printing functionality (e.g. the difference in Haskell between calling `putStrLn` and `print`, which escapes string before printing them). If you would instead view the output string in a browser, you would get the exact same output I manually unescaped above.

> After elm-format, """a\"b""" becomes """a"b"""

Yuck, I remember having [hit this before](https://github.com/avh4/elm-format/issues/175). Seems like Elm treats \\" and " the same inside multi-line strings. The solution is to escape backslashes even in multi-line strings in order for them to appear, pretty much the way `Tests.Graph.DOT` currently does it. So, `"""a\\"b"""`.

> extra layer of escaping done to it before printing it

Ah yes, exactly. See above. I'm pretty sure that we should escape strings. It's a shame that the Elm REPL doesn't allow verbatim copying, but that's just due to lack of a proper printing API. What counts is the actual String that is generated, not how it's printed on the Terminal. If you put that String in a text area, it would show up as the manually unescaped string above. Therefore, we need to properly escape it in `output*`.

Reply:
Btw., I realized that there's literally *no* `dot` implementation that only escapes " with \\. They all can't cope with nodes like (dot format) `0 [label="a\b"]` and need backslashes to be escaped, too (e.g. `0 [label="a\\b"]`). We should really find a library that can escape strings.